pkg/datamover/repository_server.go

```diff
@@ -0,0 +1,139 @@
+	"github.com/kanisterio/kanister/pkg/param"
+)
+
+// Check that RepositoryServer implements DataMover interface
+var _ DataMover = (*repositoryServer)(nil)
+
+type repositoryServer struct {
+	outputName       string
+	repositoryServer *param.RepositoryServer
+	snapJSON         string
+}
+
+func (rs *repositoryServer) Pull(ctx context.Context, sourcePath, destinationPath string) error {
+	kopiaSnap, err := rs.unmarshalKopiaSnapshot()
+	if err != nil {
+		return err
+	}
+	password, err := rs.connectToKopiaRepositoryServer(ctx)
+	if err != nil {
+		return err
+	}
+	return kopiaLocationPull(ctx, kopiaSnap.ID, destinationPath, sourcePath, password)
+}
+
+func (rs *repositoryServer) Push(ctx context.Context, sourcePath, destinationPath string) error {
+	password, err := rs.connectToKopiaRepositoryServer(ctx)
+	if err != nil {
+		return err
+	}
+	return kopiaLocationPush(ctx, destinationPath, rs.outputName, sourcePath, password)
+}
+
+func (rs *repositoryServer) Delete(ctx context.Context, destinationPath string) error {
+	kopiaSnap, err := rs.unmarshalKopiaSnapshot()
+	if err != nil {
+		return err
+	}
+	password, err := rs.connectToKopiaRepositoryServer(ctx)
+	if err != nil {
+		return err
+	}
+	return kopiaLocationDelete(ctx, kopiaSnap.ID, destinationPath, password)
+}
+
+func (rs *repositoryServer) connectToKopiaRepositoryServer(ctx context.Context) (string, error) {
+	hostname, userPassphrase, err := rs.hostnameAndUserPassphrase()
+	if err != nil {
+		return "", errors.Wrap(err, "Error Retrieving Hostname and User Passphrase from Repository Server")
+	}
+	certData, err := rs.tlsCertificateData()
+	if err != nil {
+		return "", errors.Wrap(err, "Error Retrieving TLS Certificate Data from Repository Server")
+	}
+	return userPassphrase, repository.ConnectToAPIServer(
+		ctx,
+		certData,
+		userPassphrase,
+		hostname,
+		rs.repositoryServer.Address,
+		rs.repositoryServer.Username,
+		rs.repositoryServer.ContentCacheMB,
+		rs.repositoryServer.MetadataCacheMB,
+	)
+}
+
+func (rs *repositoryServer) unmarshalKopiaSnapshot() (*snapshot.SnapshotInfo, error) {
+	if rs.snapJSON == "" {
+		return nil, errors.New("kopia snapshot information is required to delete data using kopia")
+	}
+	kopiaSnap, err := snapshot.UnmarshalKopiaSnapshot(rs.snapJSON)
+	if err != nil {
+		return nil, err
+	}
+	return &kopiaSnap, nil
+}
+
+func (rs *repositoryServer) hostnameAndUserPassphrase() (string, string, error) {
+	userCredsJSON, err := json.Marshal(rs.repositoryServer.Credentials.ServerUserAccess.Data)
+	if err != nil {
+		return "", "", errors.Wrap(err, "Error Unmarshalling User Credentials")
+	}
+	var userAccessMap map[string]string
+	if err := json.Unmarshal([]byte(string(userCredsJSON)), &userAccessMap); err != nil {
+		return "", "", errors.Wrap(err, "Failed to unmarshal User Credentials Data")
+	}
+
+	var userPassPhrase string
+	var hostName string
+	for key, val := range userAccessMap {
+		hostName = key
+		userPassPhrase = val
+	}
```

Comment:
If there are multiple entries, the hostName and userPassPhrase will always get overridden. Is it expected?

Reply:
As per the discussion, I've added functionality to pass the hostname which end user wants to use, and if it's not provided, we'll be using the first hostname and password from the map